### PR TITLE
EditPostViewController refactor (1)

### PR DESCRIPTION
**Pull request by koke**
Encapsulate all the logic of choosing what editor to launch and instantiating it inside `EditorSettings`.

I've tried to simplify the code in preparation for refactoring the EditPostViewController hierarchy, but also created a new `Restorer` class for restoring simple `UINavigationControllers` to be presented modally. There's no need to have multiple VCs with custom identifiers for that, the app delegate can take care of any VC without a `restorationClass`

The two `instantiate*` methods will get much simpler with Swift 4, but I don't think the workaround is too bad.

```swift
// Swift 4
func instantiatePostEditor(post: AbstractPost) -> UIViewController & PostEditor {
    switch editor {
    case .aztec:
        return AztecPostViewController(post: post)
    case .hybrid:
        return WPPostViewController(post: post)
    case .legacy:
        return WPLegacyEditPostViewController(post: post)
    }
}
```

First part of #7283 

To test:

For all three editors:
- Create a post
- Edit an existing post
- Create a page
- Edit an existing page
- Verify that VC restoration works

Needs review: @nheagy

### CODE REVIEW

File: WordPress/Classes/ViewRelated/Pages/PageListViewController.swift [hunk 1]
```diff
@@ -318,115 +318,40 @@ class PageListViewController: AbstractPostListViewController, UIViewControllerRe
...
-                        if let postStatus = postViewController.post.status {
-                            self?.updateFilterWithPostStatus(postStatus)
-                        }
-                    } else {
-                        self?.updateFilter(index: filterIndex)
-                    }
-                    navController.dismiss(animated: true, completion: nil)
-                }
-            } else {
-                let postViewController: EditPageViewController = EditPageViewController(draftFor: blog)
```
Comment (nheagy) [inline WordPress/Classes/ViewRelated/Pages/PageListViewController.swift]: the `draftFor:` method sets `ownsPost` within `EditPageViewController` to true, whereas `initWithPost:mode:` does not. The result is that cancelled new posts and pages now stick around as untitled drafts.

File: WordPress/Classes/Services/EditorSettings.swift [hunk 2]
```diff
@@ -67,4 +84,42 @@ class EditorSettings: NSObject {
...
+
+    // We can't return a type that's both a PostEditor and a UIViewController, so using
+    // a configure block as a hack.
+    // In Swift 4, we'll be able to do `instantiateEditor() -> UIViewController & PostEditor`,
+    // and then let the caller configure the editor.
+    func instantiatePostEditor(post: AbstractPost, configure: (PostEditor, UIViewController) -> Void) -> UIViewController {
+        switch editor {
+        case .aztec:
+            let vc = AztecPostViewController(post: post)
+            configure(vc, vc)
```
Comment (nheagy) [inline WordPress/Classes/Services/EditorSettings.swift]: `configure(vc, vc)` and `return vc` are both duplicated three times here, can that become just one call each?

File: WordPress/Classes/Services/EditorSettings.swift [hunk 3]
```diff
@@ -67,4 +84,42 @@ class EditorSettings: NSObject {
             database.set(newValue, forKey: nativeEditorEnabledKey)
         }
     }
+
+    // We can't return a type that's both a PostEditor and a UIViewController, so using
+    // a configure block as a hack.
+    // In Swift 4, we'll be able to do `instantiateEditor() -> UIViewController & PostEditor`,
+    // and then let the caller configure the editor.
+    func instantiatePostEditor(post: AbstractPost, configure: (PostEditor, UIViewController) -> Void) -> UIViewController {
```
Comment (koke) [inline WordPress/Classes/Services/EditorSettings.swift]: It feels slightly off, but then all EditorSettings does so far is figure out which editor is enabled, so I think it fits: you configure it from the settings toggles, and as a result you can instantiate an editor of the right type.

File: WordPress/Classes/Services/EditorSettings.swift (hunk 2 above)
Comment (koke) [inline WordPress/Classes/Services/EditorSettings.swift]: That's the deceptively tricky bit ;)

To do that, you'd extract the `configure`/`return` out of each individual `case` to after the `switch`. But that means you then have to declare `vc` *before* the `switch`, and what's it's type? In Swift 4, you could do:

```swift
let vc: PostEditor & UIViewController // 👈 Swift 4 feature
switch editor {
case .aztec:
    vc = AztecPostViewController(post: post)
//...
}
configure(vc, vc)
return vc
```

But then you wouldn't need the configure hack 😁

File: WordPress/Classes/ViewRelated/Pages/PageListViewController.swift (hunk 1 above)
Comment (koke) [inline WordPress/Classes/ViewRelated/Pages/PageListViewController.swift]: 😱 good catch

File: WordPress/Classes/Services/EditorSettings.swift (hunk 2 above)
Comment (nheagy) [inline WordPress/Classes/Services/EditorSettings.swift]: Ah, right. I forgot that part of if when I was thinking of the duplication.

**Comment by koke** [comment]
I refactored the shouldRemoveOnDismiss logic into a property of AbstractPost so the three editors use the same thing. @nheagy ready for another look

**Pull request merged**